### PR TITLE
Resolve #1126 -- Map system rework

### LINKED ISSUE #1126: Unhardcode Map Data
Map data such as objects to spawn and where to spawn them is present in Content, but unused.
This has caused map related scripts to hardcode this data when needed (ex: [SummonersRift.cs](https://github.com/LeagueSandbox/GameServer/blob/indev/GameServerLib/Maps/SummonersRift.cs#L252)).
Implement the content reading for the MapData, and replace hardcoded cases with it.

### PR DESCRIPTION
**Pull request by cabeca1143**
*MapData:
-Introduced ExpCurve
-Introduced Death Times
-Introduced StatsProgression
-Introduced SpawnBarracks

*MapsObject:
-Introduced GetGameObjectType() to define what type of object the file read is.
-Introduce GetTeamID() to define from What team the Object is.
-Introduced GetTeamName() to define name of the team (used mostly for models).
-Introduced GetLaneID() to define From which lane objects are.

*Package.cs
-Introduced functions to read ExpCurve, Death Times, StatsProgression and SpawnBarracks files.

*ContentManager:
-Changed GetContentFileFromJson to support a 3rd string argument, in order to input a custom subpath.

*Champion.cs:
-Unhardcoded respawn posistion.
-Unhardcoded EXP Curve.

*Summoners Rift Map
-Removed EndGameCameraPosition 
-Removed SpawnsByTeam
-Removed ExpToLevelUp
-Removed GetRespawnLocation
-Removed a bunch of lists for buildings for simply 3 lists, _nexus, _inhibitors and _turrets
-Unhardcoded the spawn for all buildings
-Unhardcoded protections
-Unhardcoded GetMinionSpawnPosition, and Spawn()

*Removed MinionsSpawnPosition.cs

*Huge thanks to Lizardy for helping me out

Resolve #1126

### CODE REVIEW

File: GameServerLib/GameObjects/AttackableUnits/AI/Champion.cs [hunk 1]
```diff
@@ -231,25 +238,21 @@ public void Recall()
...
-            if (stats.Experience < expMap[stats.Level])
-            {
-                return false;
-            }
+            var expMap = _game.Config.MapData.ExpCurve;
 
-            while (stats.Level < expMap.Count && stats.Experience >= expMap[stats.Level])
+            //Ideally we'd use "stats.Level < expMap.Count + 1", but since we still don't have gamemodes implemented yet, i'll be hardcoding the EXP level to cap at lvl 18,
+            //Since the SR Map has 30 levels in total because of URF
+            while (stats.Level < 18 && (stats.Level < 1 || stats.Experience >= expMap[stats.Level - 1])) //The + and - 1s are there because the XP files don't have level 1
```
Comment (danil179) [inline GameServerLib/GameObjects/AttackableUnits/AI/Champion.cs]: Why this loop is while if you return true inside immediately? Maybe change to `if`?

File: GameServerLib/Maps/SummonersRift.cs [hunk 2]
```diff
@@ -247,135 +312,262 @@ public void Init()
...
-                _game.ProtectionManager.AddProtection(_blueInhibTurrets[i], false, _blueInnerTurrets[i]);
-                _game.ProtectionManager.AddProtection(_blueInnerTurrets[i], false, _blueOuterTurrets[i]);
+            // Below is where we create the buildings.
+
+            // These two are used for fixing any wrongly indexed turrets that are present.
+            List<MapData.MapObject> missedTurrets = new List<MapData.MapObject>();
+
+            // This is all of the possible inhibitor NetIDs, in the order of bot-mid-top, with 0 -> 2 being Order, and 3 -> 5 being Chaos.
+            // Figure out why they hardcoded the NetIDs (and if we can work around it, do so to maximize moddability).
+            uint[] inhibitorNetIDs = new uint[6] { 0xff9303e1, 0xff4a20f1, 0xffd23c3e, 0xff26ac0f, 0xffff8f1f, 0xff6793d0 };
```
Comment (danil179) [inline GameServerLib/Maps/SummonersRift.cs]: If I remember correctly these net ids are generated from constant string (something like 0xff000000 or something similar) and the hash (there are hash functions somewhere in the project) of the inhibitor name. I'll add it as an issue.

**Review by danil179 — COMMENTED** [review]
Some comments. Hardcoded net ids in general should be related to hashed names +(bitwise OR) an offset, but I'm not sure. I'll open it as general issue.

I'll have to test it before merging.

File: GameServerLib/Maps/SummonersRift.cs [hunk 3]
```diff
@@ -635,15 +805,45 @@ public bool Spawn()
...
+                {
+                    if (barrack.Value.Name.Contains("__C"))
+                    {
+                        waypoint = RedMidWaypoints;
+                        item2 = 0xff4a20f1;
+                    }
+                    else if (barrack.Value.Name.Contains("__L"))
+                    {
+                        waypoint = RedTopWaypoints;
+                        item2 = 0xffd23c3e;
```
Comment (danil179) [inline GameServerLib/Maps/SummonersRift.cs]: I was able to verify that these net ids are generated as follows:
CRC32(original barracks name - here it is "Barracks_T1_L1") | 0xFF000000

File: GameServerLib/Maps/SummonersRift.cs [hunk 4]
```diff
@@ -635,15 +805,45 @@ public bool Spawn()
...
+                    }
+                    else if (barrack.Value.Name.Contains("__R"))
+                    {
+                        waypoint = RedBotWaypoints;
+                        item2 = 0xff9303e1;
```
Comment (danil179) [inline GameServerLib/Maps/SummonersRift.cs]: I was able to verify that these net ids are generated as follows:
CRC32(original barracks name - here it is "Barracks_T1_R1") | 0xFF000000

File: GameServerLib/Maps/SummonersRift.cs [hunk 5]
```diff
@@ -635,15 +805,45 @@ public bool Spawn()
...
+                        waypoint = BlueBotWaypoints;
+                        item2 = 0xff26ac0f;
+                    }
+                }
+                else if (barrack.Value.Name.Contains("__P_Chaos"))
+                {
+                    if (barrack.Value.Name.Contains("__C"))
+                    {
+                        waypoint = RedMidWaypoints;
+                        item2 = 0xff4a20f1;
```
Comment (danil179) [inline GameServerLib/Maps/SummonersRift.cs]: I was able to verify that these net ids are generated as follows:
CRC32(original barracks name - here it is "Barracks_T1_C1") | 0xFF000000

**Review by danil179 — COMMENTED** [review]
Some comments about net ids generation.

File: GameServerLib/Maps/SummonersRift.cs (hunk 2 above)
Comment (danil179) [reply GameServerLib/Maps/SummonersRift.cs]: These net ids are generated as follows:
CRC32(original barracks name - for example "Barracks_T1_L1","Barracks_T2_C1","Barracks_T1_R1") | 0xFF000000

See BaseTurret.cs for the CRC32 lib in C# that we are already using.

File: GameServerLib/Maps/SummonersRift.cs [hunk 6]
```diff
@@ -805,7 +794,7 @@ public bool Spawn()
                 new Tuple<long, int>(35 * 60 * 1000, 0)
             };
 
-            var spawnToWaypoints = new Dictionary<string, Tuple<List<Vector2>, uint>>();
+            /*var spawnToWaypoints = new Dictionary<string, Tuple<List<Vector2>, uint>>();
```
Comment (danil179) [inline GameServerLib/Maps/SummonersRift.cs]: Remove commented code.

**Review by danil179 — COMMENTED** [review]
Remove commented code. I'll also test it in the next hour.

File: GameServerLib/Maps/SummonersRift.cs (hunk 6 above)
Comment (cabeca1143) [reply GameServerLib/Maps/SummonersRift.cs]: Resolved

**Review by danil179 — APPROVED** [review]
Approved. Will test and merge.

**Pull request merged**